### ISSUE TITLE
Deleting the Instance ID token does not trigger onNewToken

### ISSUE THREAD

**Issue by helenaford** [firebase/firebase-android-sdk#2485]
### [READ] Step 1: Are you in the right place?
Yes

### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 4.1.1
  * Firebase Component: Messaging 
  * Component version:  21.0.1

### [REQUIRED] Step 3: Describe the problem
Inconsistency between iOS and Android SDKs when a token is deleted. On iOS, there was an update to the SDK to make sure the listener refreshed, please see p/r https://github.com/firebase/firebase-ios-sdk/pull/5339. Unable to reproduce same behaviour on Android. 

#### Steps to reproduce:

1. get Firebase Messaging Token `FirebaseInstanceId.getInstance().getToken();`
2. delete Firebase Messaging Token `FirebaseInstanceId.getInstanceId().deleteInstanceId()`
3. get Firebase Messaging Token `FirebaseInstanceId.getInstance().getToken();`

Token is updated, but `FirebaseMessagingService onNewToken ` is not triggered, unlike on iOS.

**Comment by ciarand** [comment]
Hi, thanks for this report!

We definitely do want to trigger the `onNewToken` when a new token is generated (i.e. after steps 1 and 3). I did some tests and wasn't able to reproduce this with the no-arg `FirebaseInstanceId#getToken()` call. At least locally it seems like the `onNewToken` hook does end up getting called after the background token creation finishes.

Was this in a test case? How long did you wait after step 3? The blocking / nonblocking behavior is definitely unintuitive and something we're working on improving in the future.

That being said, the version that takes an explicit sender ID (`FirebaseInstanceId#getToken(String, String)`) as well as the `getInstanceId()` methods do not invoke that hook which is definitely a bug. I'm working with the rest of the team on the exact release timing, but I'm hoping we'll be able to have a fix for that out shortly.

**Comment by helenaford** [comment]
Hi @ciarand, no problem. thanks for taking the time to look into it! I found the bug via `react-native-firebase` and also reported it there (https://github.com/invertase/react-native-firebase/issues/4997). There may be some sort of 'patch' that can be put in place for `RNFirebase`, but of course, this won't fix other packages like `flutterfire` which it can be replicated in that too.

Not sure how time is relevant, maybe it is? But, I've been waiting over 1 minute and no event comes through. I was unable to go deeper through the codebase as it looks like the FirebaseInstanceId package is private for android.

When you said you can't reproduce this issue with the no-arg `FirebaseInstanceId#getToken()`, that's interesting. Both packages I mentioned pass in the args ([`RNFirebase`]( https://github.com/invertase/react-native-firebase/blob/a61d0cdebb6c2c7c9c3966f0a7a0af6a0ef3ae05/packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingModule.java#L116) and [FlutterFire](https://github.com/FirebaseExtended/flutterfire/blob/826c4a315aaf4852fb4a4aabb5ab34136ed07cab/packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java#L166)). 

But, I've seen that `FirebaseInstanceId` is deprecated, and that either `FirebaseMessaging.getToken()` or `FirebaseInstallations.getToken(boolean forceRefresh)` should be used instead, taken from the [docs](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceId):

`Firebase Instance ID has been replaced with FirebaseInstallations for app instance identifiers and FirebaseMessaging.getToken() for FCM registration tokens.`


If the packages updated to use `FirebaseMessaging.getToken()`, would that make any difference?

**Comment by ciarand** [comment]
I've put together a fix for this (internal ref 365888536) and it should be rolling out in the next FCM release.

>But, I've seen that FirebaseInstanceId is deprecated, and that either FirebaseMessaging.getToken() or FirebaseInstallations.getToken(boolean forceRefresh) should be used instead, taken from the docs:

Yes, in general the future is to go through either `FirebaseInstallations` (for IDs) or `FirebaseMessaging` (for FCM registration tokens). The `FirebaseInstallations.getToken` method returns an installations auth token, which is distinct from the FCM registration token. For sending an FCM message to an app instance, the `FirebaseMessaging.getToken` method is the right one to use.

**Comment by helenaford** [comment]
Thank you, look forward to the fix! Do have a rough eta?

**Comment by ciarand** [comment]
Apologies, I missed this before. This should be live now: https://firebase.google.com/support/release-notes/android#iid_v21-1-0

**Comment by mikehardy** [comment]
@ciarand great! @helenaford note this is packaged in firebase-android-sdk 27.0.0 which is a breaking change for react-native-firebase and we haven't ported yet :sweat_smile: https://github.com/invertase/react-native-firebase/issues/5150

I believe based on BoM comparision if you are not using the admob module that you can overrde the sdk and everything will be fine but it's untested as of yet
